commitrmnocb/outcome.go

```diff
@@ -0,0 +1,316 @@
+package commitrmnocb
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"golang.org/x/exp/maps"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+)
+
+// Outcome depending on the current state, either:
+// - chooses the seq num ranges for the next round
+// - builds a report
+// - checks for the transmission of a previous report
+func (p *Plugin) Outcome(
+	outCtx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation,
+) (ocr3types.Outcome, error) {
+	previousOutcome, nextState := p.decodeOutcome(outCtx.PreviousOutcome)
+	commitQuery := Query{}
+
+	consensusObservation, err := p.getConsensusObservation(aos)
+	if err != nil {
+		return ocr3types.Outcome{}, err
+	}
+
+	outcome := Outcome{}
+
+	switch nextState {
+	case SelectingRangesForReport:
+		outcome = p.ReportRangesOutcome(commitQuery, consensusObservation)
+
+	case BuildingReport:
+		outcome = p.buildReport(commitQuery, consensusObservation)
+
+	case WaitingForReportTransmission:
+		outcome = p.checkForReportTransmission(previousOutcome, consensusObservation)
+
+	default:
+		p.lggr.Warnw("Unexpected state in Outcome", "state", nextState)
+		return outcome.Encode()
+	}
+
+	p.lggr.Infow("Commit Plugin Outcome", "outcome", outcome, "oid", p.nodeID)
+	return outcome.Encode()
+}
+
+// ReportRangesOutcome determines the sequence number ranges for each chain to build a report from in the next round
+func (p *Plugin) ReportRangesOutcome(
+	query Query,
+	consensusObservation ConsensusObservation,
+) Outcome {
+	rangesToReport := make([]ChainRange, 0)
+
+	rmnOnRampMaxSeqNumsMap := make(map[cciptypes.ChainSelector]cciptypes.SeqNum)
+	for _, seqNumChain := range query.RmnOnRampMaxSeqNums {
+		rmnOnRampMaxSeqNumsMap[seqNumChain.ChainSel] = seqNumChain.SeqNum
+	}
+
+	observedOnRampMaxSeqNumsMap := consensusObservation.OnRampMaxSeqNums
+	observedOffRampNextSeqNumsMap := consensusObservation.OffRampNextSeqNums
+
+	for chainSel, offRampNextSeqNum := range observedOffRampNextSeqNumsMap {
+		onRampMaxSeqNum, exists := observedOnRampMaxSeqNumsMap[chainSel]
+		if !exists {
+			continue
+		}
+
+		if rmnOnRampMaxSeqNum, exists := rmnOnRampMaxSeqNumsMap[chainSel]; exists {
+			onRampMaxSeqNum = min(onRampMaxSeqNum, rmnOnRampMaxSeqNum)
+		}
+
+		if offRampNextSeqNum <= onRampMaxSeqNum {
+			chainRange := ChainRange{
+				ChainSel:    chainSel,
+				SeqNumRange: [2]cciptypes.SeqNum{offRampNextSeqNum, onRampMaxSeqNum},
+			}
+			rangesToReport = append(rangesToReport, chainRange)
+		}
+	}
+
+	outcome := Outcome{
+		OutcomeType:             ReportIntervalsSelected,
+		RangesSelectedForReport: rangesToReport,
+	}
+
+	return outcome
+}
+
+// Given a set of observed merkle roots, gas prices and token prices, and roots from RMN, construct a report
+// to transmit on-chain
+func (p *Plugin) buildReport(
```

Comment:
nit: this is a pure function
```suggestion
// buildReport takes observed merkle roots, gas prices and token prices, and roots from RMN, construct a report
// to transmit on-chain
func buildReport(
```

Reply:
I'll leave out any further comments like this. I'd be tempted to make most of the helpers in this file pure functions by surfacing the logger dependency (or logging errors outside the helper).